examples/sealed-auctions/src/types.rs

```diff
@@ -0,0 +1,83 @@
+extern crate chrono;
+
+use std::{collections::HashSet, vec::Vec}; // `use` statements can go anywhere
+
+pub type UserId = mantle::types::Address;
+pub type ItemId = usize;
+
+// A bid struct with the bidder and latest bid
+#[derive(Serialize, Deserialize, Clone, Debug, Eq)]
+pub struct Bid {
+    pub bidder: UserId,
+    pub value: usize,
+}
+
+impl PartialEq for Bid {
+    fn eq(&self, other: &Bid) -> bool {
+        self.bidder == other.bidder
+    }
+}
+
+impl PartialOrd for Bid {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.value.partial_cmp(&other.value)
+    }
+}
+
+impl std::hash::Hash for Bid {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.bidder.hash(state);
+    }
+}
+
+impl std::borrow::Borrow<UserId> for Bid {
+    fn borrow(&self) -> &UserId {
+        &self.bidder
+    }
+}
+
+// The auction struct for every auction that has taken place on the platform
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct Auction {
+    pub item_id: ItemId,
+    pub bids: HashSet<Bid>,
+    pub reserve: usize,
+    pub max_bid: usize,
+    pub realized: usize,
+    pub seller: UserId,
+    pub buyer: UserId,
+}
+
+// We want to emit an event when a new item comes into the auction market
+// but don't want to expose anything in the event other than the itemID
+// and description. Hence, this struct carries `Event`.
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Event)]
```

Comment:
You know, this is a really cool use of `Event`! I had never thought that people would include events as part of their state. In fact, I was considering making `fn emit(self)` instead of `fn emit(&self)`, but clearly the former is more flexible. Nice find!

Reply:
Thank you Sir!